### PR TITLE
Launch dev channel deprecation infobar at startup

### PR DESCRIPTION
**Pull request by simonhong**
Resolves https://github.com/brave/brave-browser/issues/33822

image

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Launch dev-channel release and check infobar is launched at startup
2. Click x button to close and relaunch and check infobar is shown again

### CODE REVIEW

File: app/brave_generated_resources.grd [hunk 1]
```diff
@@ -456,7 +456,10 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
           With a Directory
         </message>
       </if>
-
+      <!-- Dev channel deprecation infobar -->
+      <message name="IDS_DEV_CHANNEL_DEPRECATION_INFOBAR_MESSAGE" desc="The text label of dev channel deprecation announcement.">
+        Beginning on December 1, 2023, Brave Dev channel will no longer receive browser updates. We recommend users migrate to Nightly, Beta, or Stable versions of Brave.
```
[the remark this answers is not in the document]
Comment (simonhong) [reply app/brave_generated_resources.grd:461]: > We could potentially steer people towards Beta specifically, as this is the channel that had the exact same content as Dev for a long time. Not a big deal, something to consider.
> 
> Also, the stable channel is called Release, so we might want to stick to that terminology. Again, not a big deal, something to consider.
> 
> ~I assume the "Learn more" link will point to the community post about this? Is that up already?~ I see the post is up and will be linked to.

Agreed with @wknapik 's comment. @rebron WDYT?

**Review by sangwoo108 — APPROVED** [review]
LGTM👍

File: chromium_src/chrome/browser/ui/startup/infobar_utils.cc [hunk 2]
```diff
@@ -33,5 +35,35 @@ class BraveGoogleKeysInfoBarDelegate {
...
+      !IsAutomationEnabled()) {
+    static bool infobars_shown = false;
+    if (infobars_shown) {
+      return;
+    }
+    infobars_shown = true;
+
+    content::WebContents* web_contents =
+        browser->tab_strip_model()->GetActiveWebContents();
+    DCHECK(web_contents);
```
Comment (goodov) [inline chromium_src/chrome/browser/ui/startup/infobar_utils.cc:64]: is this really true? I see this check in the ChromiumImpl:
```
  if (!browser || !profile || browser->tab_strip_model()->count() == 0)
    return;
```

Comment (simonhong) [reply chromium_src/chrome/browser/ui/startup/infobar_utils.cc:68]: Yeah, added this checking. thanks 👍🏼

**Review by goodov — APPROVED** [review]

File: app/brave_generated_resources.grd [hunk 3]
```diff
@@ -462,7 +462,10 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
...
```
Comment (bsclifton) [inline app/brave_generated_resources.grd:467]:
```suggestion
        Beginning on December 1, 2023, Brave Dev channel will no longer receive browser updates. We recommend users migrate to the Nightly, Beta, or Release versions of Brave.
```

Comment (simonhong) [reply app/brave_generated_resources.grd]: Updated! 👍🏼

**Review by bsclifton — APPROVED** [review]

**Pull request merged**